components/script/dom/globalscope.rs

```diff
@@ -3149,9 +3148,16 @@ impl GlobalScope {
                             gamepad.update_connected(true, has_gesture);
                         }
                     }
-                    let gamepad_list = window.Navigator().gamepads();
-                    let gamepad_arr: [DomRoot<Gamepad>; 1] = [gamepad.clone()];
-                    gamepad_list.add_if_not_exists(&gamepad_arr);
+                    let navigator = window.Navigator();
+                    let mut gamepad_list = navigator.gamepads();
+                    if !gamepad_list
+                        .iter()
+                        .any(|g| g.as_ref().map_or(false, |g| g.gamepad_id() == gamepad.gamepad_id()))
+                    {
+                        // Ensure that the gamepad has the correct index
+                        gamepad.update_index(gamepad_list.len() as i32);
```

Comment:
I don't think this is correct, because the index should not change. Instead, you should fill any missing indices in the list with `None`. This could happen if the focused webview changes. 

Btw, in webview, you can nest the index creation inside the `EventType::Connected` variant handling, and you can have a single `self.event_queue.push(EmbedderEvent::Gamepad(event));` for all variants.

Reply:
I originally moved that out from the old GamepadList class, but I've actually now updated it with the more up-to-date spec steps on selecting an unused index, which will be in my next set of changes.